### PR TITLE
Sequential run of backfilled attempts

### PR DESCRIPTION
**Pull request by frsyuki**

### CODE REVIEW

File: digdag-core/src/main/java/io/digdag/core/database/migrate/Migration_20161209001857_CreateDelayedSessionAttempts.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+import org.skife.jdbi.v2.Handle;
+
+public class Migration_20161209001857_CreateDelayedSessionAttempts
+        implements Migration
+{
+    @Override
+    public void migrate(Handle handle, MigrationContext context)
+    {
+        handle.update(
+                context.newCreateTableBuilder("delayed_sssion_attempts")
```
Comment (danielnorberg) [inline digdag-core/src/main/java/io/digdag/core/database/migrate/Migration_20161209001857_CreateDelayedSessionAttempts.java]: s/sssion/session

**Comment by frsyuki** [comment]
@komamitsu Can I ask you to review?
As future changes, we may need following things:

* `Optional<Long> io.digdag.client.api.RestSessionAttempt.getBlockedBySessionId()` or something similar so that digdag-cli and digdag-ui can let users know that the sessions are actually not running but blocked.
* Additional option or directive in workflow definition that lets users decide whether backfilled sessions should run concurrently or sequentially. In this change, all backfilled sessions run sequentially which seems like a better default behavior.

File: digdag-core/src/main/java/io/digdag/core/database/DatabaseSessionStoreManager.java [hunk 2]
```diff
@@ -609,6 +615,47 @@ private String dumpTaskArchive(List<ArchivedTask> tasks)
...
+    {
+        List<RuntimeException> exceptions = transaction((handle, dao) -> {
+            List<StoredDelayedSessionAttempt> locked = handle.createQuery(
+                    "select da.* from delayed_session_attempts da" +
+                    " where not exists (" +
+                        " select * from session_attempts sa" +
+                        " where sa.session_id = da.dependent_session_id" +
+                        " and " + bitAnd("sa.state_flags", Integer.toString(AttemptStateFlags.DONE_CODE)) + " = 0" +
+                    ")" +
+                    " and next_run_time <= :currentTime" +
```
Comment (komamitsu) [inline digdag-core/src/main/java/io/digdag/core/database/DatabaseSessionStoreManager.java]: `order by id` isn't needed here?

File: digdag-core/src/main/java/io/digdag/core/database/DatabaseSessionStoreManager.java [hunk 3]
```diff
@@ -609,6 +615,47 @@ private String dumpTaskArchive(List<ArchivedTask> tasks)
...
+                .mapTo(StoredDelayedSessionAttempt.class)
+                .list();
+            return locked.stream()
+                .map(delayedAttempt -> {
+                    try {
+                        func.submit(new DatabaseDelayedAttemptControlStore(handle), delayedAttempt);
+                        return null;
+                    }
+                    catch (RuntimeException ex) {
+                        return ex;
```
Comment (komamitsu) [inline digdag-core/src/main/java/io/digdag/core/database/DatabaseSessionStoreManager.java]: How about logging this exception not just filtering out it?

File: digdag-core/src/main/java/io/digdag/core/database/DatabaseSessionStoreManager.java [hunk 4]
```diff
@@ -1328,6 +1465,14 @@ public void insertMonitors(long attemptId, List<SessionMonitor> monitors)
                 " values (:projectId, :workflowName, :sessionTime)")
         void upsertAndLockSession(@Bind("projectId") int projectId,
                 @Bind("workflowName") String workflowName, @Bind("sessionTime") long sessionTime);
+
+        @SqlQuery("select id from sessions s" +
+                " where id = (" +
+                    "select session_id from session_attempts" +
+                    " where id = :attemptId" +
+                ")" +
+                " for update")
```
Comment (komamitsu) [inline digdag-core/src/main/java/io/digdag/core/database/DatabaseSessionStoreManager.java]: Using `EXISTS` would be better?
```
select id from sessions s
where exists (
  select session_id from session_attempts
  where id = :attemptId and session_id = s.id
)
```

Comment (komamitsu) [inline digdag-core/src/main/java/io/digdag/core/database/DatabaseSessionStoreManager.java]: Ah, there is no significant difference between them... NVM

File: digdag-core/src/main/java/io/digdag/core/schedule/ScheduleExecutor.java [hunk 5]
```diff
@@ -335,41 +377,97 @@ private Scheduler getSchedulerOfSchedule(StoredSchedule sched)
...
+        }
+        catch (ResourceLimitExceededException ex) {
+            logger.warn("Failed to start delayed attempt Due to too many active tasks. Will be retried after 5 minutes.", ex);
+            control.delayDelayedAttempt(delayedAttempt.getAttemptId(), Instant.now().plusSeconds(5 * 60));
+            return;
+        }
+        control.completeDelayedAttempt(delayedAttempt.getAttemptId());
+    }
+
+    private AttemptRequest newAttemptrequest(
```
Comment (komamitsu) [inline digdag-core/src/main/java/io/digdag/core/schedule/ScheduleExecutor.java]: `s/newAttemptrequest/newAttemptRequest/`

File: digdag-core/src/main/java/io/digdag/core/schedule/ScheduleExecutor.java [hunk 6]
```diff
@@ -335,41 +377,97 @@ private Scheduler getSchedulerOfSchedule(StoredSchedule sched)
...
-                        );
-                }
-                else {
-                    try {
-                        StoredSessionAttemptWithSession attempt = handler.start(def,
-                                ScheduleTime.of(instant, sched.getNextScheduleTime()),
-                                Optional.of(attemptName), sched.getLastSessionTime());
-                        attempts.add(attempt);
-                        // TODO this may throw ResourceLimitExceededException. But some sessions are already committed. To be able to rollback everything, inserting all sessions needs to be in a single transaction.
+            return workflowExecutor.submitTransaction(siteId, (submitter) -> {
```
Comment (komamitsu) [inline digdag-core/src/main/java/io/digdag/core/schedule/ScheduleExecutor.java]: The above line https://github.com/treasure-data/digdag/pull/431/files#diff-a0e8a0113728f65511442e76d9e406adR339 can be removed

```
            ScheduleControl lockedSched = new ScheduleControl(store, sched);
```

File: digdag-core/src/main/java/io/digdag/core/database/DatabaseSessionStoreManager.java (hunk 4 above)
Comment (komamitsu) [inline digdag-core/src/main/java/io/digdag/core/database/DatabaseSessionStoreManager.java]: `upsertAndLockSession()` uses a different condition to do the same thing? If so, we'd better use the same condition in both methods for better code maintenance.

**Comment by komamitsu** [comment]
@frsyuki Can you take a look at this comment? https://github.com/treasure-data/digdag/pull/431/files#r95534579

Other than that look good although I left some other minor comments.

File: digdag-core/src/main/java/io/digdag/core/schedule/ScheduleExecutor.java (hunk 6 above)
Comment (frsyuki) [inline digdag-core/src/main/java/io/digdag/core/schedule/ScheduleExecutor.java]: 👍

File: digdag-core/src/main/java/io/digdag/core/database/DatabaseSessionStoreManager.java (hunk 2 above)
Comment (frsyuki) [inline digdag-core/src/main/java/io/digdag/core/database/DatabaseSessionStoreManager.java]: `order by next_run_time asc`?

File: digdag-core/src/main/java/io/digdag/core/database/DatabaseSessionStoreManager.java (hunk 4 above)
Comment (frsyuki) [inline digdag-core/src/main/java/io/digdag/core/database/DatabaseSessionStoreManager.java]: H2 can't use JOIN with FOR UPDATE. So, H2Dao.upsertAndLockSession returns only `id`.
PostgreSQL can use JOIN and can return more columns together (sa.* and s.*).

File: digdag-core/src/main/java/io/digdag/core/schedule/ScheduleExecutor.java (hunk 5 above)
Comment (frsyuki) [inline digdag-core/src/main/java/io/digdag/core/schedule/ScheduleExecutor.java]: 👍

File: digdag-core/src/main/java/io/digdag/core/database/DatabaseSessionStoreManager.java (hunk 4 above)
Comment (komamitsu) [inline digdag-core/src/main/java/io/digdag/core/database/DatabaseSessionStoreManager.java]: 👍

**Comment by komamitsu** [comment]
👍

**Pull request merged**